msm/src/circuit_design/composition.rs

```diff
@@ -0,0 +1,249 @@
+/// Tools to compose different circit designers.
+use crate::{
+    circuit_design::capabilities::{ColAccessCap, ColWriteCap, LookupCap},
+    columns::ColumnIndexer,
+    logup::LookupTableID,
+};
+use ark_ff::PrimeField;
+
+/// `MPrism` allows one to Something like a Prism, but for Maybe and not just any Applicative.
+///
+/// See
+/// - <https://hackage.haskell.org/package/lens-4.17.1/docs/Control-Lens-Prism.html>
+/// - <https://hackage.haskell.org/package/lens-tutorial-1.0.4/docs/Control-Lens-Tutorial.html>
+pub trait MPrism {
+    /// The lens source type, i.e., the object containing the field.
+    type Source;
+
+    /// The lens target type, i.e., the field to be accessed or modified.
+    type Target;
+
+    fn traverse(&self, source: Self::Source) -> Option<Self::Target>;
+
+    fn re_get(&self, target: Self::Target) -> Self::Source;
+}
+
+pub struct ComposedMPrism<LHS, RHS> {
+    /// The left-hand side of the composition.
+    lhs: LHS,
+
+    /// The right-hand side of the composition.
+    rhs: RHS,
+}
+
+impl<LHS, RHS> ComposedMPrism<LHS, RHS>
+where
+    LHS: MPrism,
+    LHS::Target: 'static,
+    RHS: MPrism<Source = LHS::Target>,
+{
+    pub fn compose(lhs: LHS, rhs: RHS) -> ComposedMPrism<LHS, RHS> {
+        ComposedMPrism { lhs, rhs }
+    }
+}
+
+impl<LHS, RHS> MPrism for ComposedMPrism<LHS, RHS>
+where
+    LHS: MPrism,
+    LHS::Target: 'static,
+    RHS: MPrism<Source = LHS::Target>,
+{
+    type Source = LHS::Source;
+    type Target = RHS::Target;
+
+    fn traverse(&self, source: Self::Source) -> Option<Self::Target> {
+        let r1: Option<_> = self.lhs.traverse(source);
+        let r2: Option<_> = r1.and_then(|x| self.rhs.traverse(x));
+        r2
+    }
+
+    fn re_get(&self, target: Self::Target) -> Self::Source {
+        self.lhs.re_get(self.rhs.re_get(target))
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////
+// Interpreter and sub-interpreter
+////////////////////////////////////////////////////////////////////////////
+
+// Generic sub-environment struct. Internal object to avoid copy-paste.
+struct SubEnv<'a, F: PrimeField, CIx1: ColumnIndexer, Env1: ColAccessCap<F, CIx1>, L> {
+    env: &'a mut Env1,
+    lens: L,
+    phantom: core::marker::PhantomData<(F, CIx1)>,
+}
+
+/// Sub environment with a lens that is mapping lookup tables.
+/// Can't use `SubEnv` directly because rust is not idris.
+pub struct SubEnvColumn<'a, F: PrimeField, CIx1: ColumnIndexer, Env1: ColAccessCap<F, CIx1>, L>(
+    SubEnv<'a, F, CIx1, Env1, L>,
+);
+
+/// Sub environment with a lens that is mapping lookup tables.
+pub struct SubEnvLookup<'a, F: PrimeField, CIx1: ColumnIndexer, Env1: ColAccessCap<F, CIx1>, L>(
```

Comment:
Yes :) Rust is stupid -- ideally I would use just `SubEnv`, but because of trait limitations I can only capture one type of "lens" in one type of `SubEnv`. So there are two `SubEnv`s that are the same on the data level, but have different trait implementations.